src/sqlfluff/dialects/dialect_clickhouse.py

```diff
@@ -296,6 +296,52 @@ class UnorderedSelectStatementSegment(ansi.UnorderedSelectStatementSegment):
     )
 
 
+class OrderByClauseSegment(ansi.OrderByClauseSegment):
+    """Enhance A `ORDER BY` state to include WITH FILL.
+
+    https://clickhouse.com/docs/en/sql-reference/statements/select/order-by#order-by-expr-with-fill-modifier
+    """
+
+    match_grammar: Matchable = Sequence(
+        "ORDER",
+        "BY",
+        Indent,
+        Delimited(
+            Sequence(
+                OneOf(
+                    Ref("ColumnReferenceSegment"),
+                    # Can `ORDER BY 1`
+                    Ref("NumericLiteralSegment"),
+                    # Can order by an expression
+                    Ref("ExpressionSegment"),
+                ),
+                OneOf("ASC", "DESC", optional=True),
+                # NB: This isn't really ANSI, and isn't supported in Mysql, but
+                # is supported in enough other dialects for it to make sense here
+                # for now.
+                Sequence("NULLS", OneOf("FIRST", "LAST"), optional=True),
+                Sequence(
+                    "WITH",
+                    "FILL",
+                    Sequence("FROM", Ref("ExpressionSegment"), optional=True),
+                    Sequence("TO", Ref("ExpressionSegment"), optional=True),
+                    Sequence(
+                        "STEP",
+                        OneOf(
+                            Ref("NumericLiteralSegment"),
+                            Ref("IntervalExpressionSegment"),
+                        ),
+                        optional=True,
+                    ),
+                    optional=True,
+                ),
```

Comment:
Sorry for the thrash, but because we are copying so much of `OrderByClauseSegment`, I think it would be better to define a `WithFillSegment` that is `Nothing()` in ansi but defined in clickhouse.

So you could add `WithFillSegment` after this line: https://github.com/sqlfluff/sqlfluff/blob/60e3a34dfbc47dd193fec429bfe053d23e1481a0/src/sqlfluff/dialects/dialect_ansi.py#L2407

And in ansi it's similar to https://github.com/sqlfluff/sqlfluff/blob/60e3a34dfbc47dd193fec429bfe053d23e1481a0/src/sqlfluff/dialects/dialect_ansi.py#L776-L783

But it's overwritten in clickhouse.

What do you think?

Reply:
No stress. I'm wondering if this is the appropriate way to go about it, as we'd be leaking a non ansi specific implementation term `WithFill` into the `ansi` namespace. I get the idea and am happy to do it if you think it's best - it just seems like a pollution of the `ansi` namespace though.

Reply:
I think it's totally fine to put non-ansi `Nothing()` stuff in ansi because it doesn't change how ansi is parsed but it allows dialects to override a small part and still stay in sync with ansi as `OrderByClauseSegment` evolves over time.